gravitee-am-resource/gravitee-am-resource-infobip/src/main/java/io/gravitee/am/resource/infobip/provider/InfobipResourceProvider.java

```diff
@@ -0,0 +1,120 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.am.resource.infobip.provider;
+
+import com.infobip.ApiClient;
+import com.infobip.api.TfaApi;
+import com.infobip.model.TfaStartAuthenticationRequest;
+import com.infobip.model.TfaStartAuthenticationResponse;
+import com.infobip.model.TfaVerifyPinRequest;
+import com.infobip.model.TfaVerifyPinResponse;
+import io.gravitee.am.common.exception.mfa.InvalidCodeException;
+import io.gravitee.am.common.exception.mfa.SendChallengeException;
+import io.gravitee.am.resource.api.mfa.MFAChallenge;
+import io.gravitee.am.resource.api.mfa.MFALink;
+import io.gravitee.am.resource.api.mfa.MFAResourceProvider;
+import io.gravitee.am.resource.infobip.InfobipResourceConfiguration;
+import io.reactivex.Completable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+
+/**
+ * @author Ruan Ferreira (ruan@incentive.me)
+ * @author Incentive.me
+ */
+public class InfobipResourceProvider implements MFAResourceProvider {
+    private static final Logger LOGGER = LoggerFactory.getLogger(InfobipResourceProvider.class);
+
+    TfaApi tfaApi;
+    public static final String APPROVED = "approved";
+
+    private String pinId;
+    private String to;
+
+    @Autowired
+    private InfobipResourceConfiguration configuration;
+
+    @Override
+    public Completable send(MFALink target) {
+        this.to = target.getTarget();
+
+        String baseUrl = this.configuration.getBaseUrl();
+        String apiKey = this.configuration.getApiKey();
+        String apiKeyPrefix = this.configuration.getApiKeyPrefix();
+        String messageId = this.configuration.getMessageId();
+        String applicationId = this.configuration.getApplicationId();
+
+        ApiClient apiClient = new ApiClient();
+        apiClient.setApiKeyPrefix(apiKeyPrefix);
+        apiClient.setApiKey(apiKey);
+        apiClient.setBasePath(baseUrl);
+
+        this.tfaApi = new TfaApi(apiClient);
+
+        return Completable.create((emitter) -> {
+            try{
+                TfaStartAuthenticationResponse sendCodeResponse = this.tfaApi.sendTfaPinCodeOverSms(true,
+                        new TfaStartAuthenticationRequest()
+                                .applicationId(applicationId)
+                                .messageId(messageId)
+                                .from("InfoSMS")
+                                .to(this.to)
+                );
+
+                boolean isSuccessful = sendCodeResponse.getSmsStatus().equals("MESSAGE_SENT");
+
+                if(!isSuccessful) {
+                    emitter.onError(new SendChallengeException("Message not sent"));
+                }
+
+                this.pinId = sendCodeResponse.getPinId();
+
+                LOGGER.debug("Infobip Verification code asked with ID '{}'", sendCodeResponse.getPinId());
+                emitter.onComplete();
```

Comment:
Put this code in a "else" section, otherwise both emitter.onError and  emitter.onComplete(); will be called if there is an error.